### PR TITLE
[WIP] Add patience argument to run_language_modeling script

### PR DESCRIPTION
**Pull request by thesamuel**
# Summary
Often, we want to stop training if loss does not improve for a number of epochs. This PR adds a "patience" argument, which is a limit on the number of times we can get a non-improving eval loss before stopping training early.

It is implemented by other NLP frameworks, such as AllenNLP (see [trainer.py](https://github.com/allenai/allennlp/blob/master/allennlp/training/trainer.py#L95) and [metric_tracker.py](https://github.com/allenai/allennlp/blob/1a8a12cd1b065d74fec3d2e80105a684736ff709/allennlp/training/metric_tracker.py#L6)).

# Motivation
This feature allows faster fine-tuning by breaking the training loop early and avoids users the toil of checking metrics on Tensorboard.

# Caveats
Often, models are evaluated once per epoch, but run_lm_finetuning.py has an option to evaluate after a set number of model update steps (dictated by `--logging_steps` if `--evaluate_during_training` is true). Because of this, I've elected to tie patience to the number of evaluations without improvement in loss.

# To-do
- Add tests
- Fix long lines

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
This is really cool!

**Review by julien-c — APPROVED** [review]
I feel like at this point, this should be inside a Trainer class similar to what @jplu is doing for TF (shared between scripts) (and similar to what @srush is doing w/ Lightning), but in the meantime, LGTM. Thanks!

**Comment by thesamuel** [comment]
> I feel like at this point, this should be inside a Trainer class similar to what @jplu is doing for TF (shared between scripts) (and similar to what @srush is doing w/ Lightning), but in the meantime, LGTM. Thanks!

Sounds great! I'll go ahead and fix the code quality check.

**Comment by thesamuel** [comment]
Since `run_langauge_modeling.py` now uses the `Trainer` class, I'll likely create a new PR that adds perplexity to `Trainer`.
